### ISSUE TITLE
How the grad is transferred between layer

### ISSUE THREAD

**Issue by ShawnXuan** [pytorch/pytorch#2384]
consider a simple example here:
```python
import torch
from torch.autograd import Variable

input = Variable(torch.randn(20, 3, 28, 28), requires_grad=True)
m = torch.nn.Conv2d(3, 16, 5)
output = m(input)

loss = torch.sum(output)# define loss to perform backprop
m.zero_grad()
loss.backward()

print(type(input))
print(input.grad.size())
print(type(output))
print(output.grad)
```
the output is:
```
<class 'torch.autograd.variable.Variable'>
torch.Size([20, 3, 28, 28])
<class 'torch.autograd.variable.Variable'>
None
```
I find the `output.grad` is `None`. I don't know how the `input.grad` is calculated without `output.grad`.
and want to know how to get the values of `output.grad`.

thanks!

**Comment by albanD** [comment]
Hi,

By default, only the Variables that you have created with `requires_grad=True` will have their gradient saved in the `.grad` field. For all other Variables, for memory efficiency, the gradient is not saved.
If you want to save the gradient of your output to be able to inspect it, you can add a `output.retain_grad()` before calling `loss.backward()` and the `.grad` field will be populated as you expect.

**Comment by ShawnXuan** [comment]
Thanks!
`output.retain_grad()` works by upgrade to `0.2.0.post1`
